File: deeptech_semantic_search/services/embedding_service.py
"""
Embedding service using sentence-transformers
Handles AI model loading and embedding generation
"""

import numpy as np
from typing import List, Dict, Any
import asyncio
import logging

logger = logging.getLogger(__name__)

class EmbeddingService:
    def __init__(self):
        self.model = None
        self.model_name = "all-MiniLM-L6-v2"  # 384-dimensional embeddings
        self.is_initialized = False

    async def initialize(self):
        """Load the sentence transformer model"""
        try:
            print(f"🔄 Loading sentence transformer model: {self.model_name}")
            # Import here to avoid module-level import issues
            from sentence_transformers import SentenceTransformer

            print("🔄 Creating SentenceTransformer instance...")
            # Run model loading in thread pool to avoid blocking
            loop = asyncio.get_event_loop()
            self.model = await loop.run_in_executor(None, SentenceTransformer, self.model_name)
            self.is_initialized = True
            print("✅ Model loaded successfully")
        except Exception as e:
            print(f"❌ Failed to load model: {str(e)}")
            import traceback
            traceback.print_exc()
            raise

    def is_ready(self) -> bool:
        """Check if the service is ready"""
        return self.is_initialized and self.model is not None

    async def generate_embedding(self, text: str) -> List[float]:
        """
        Generate embedding for input text
        Returns 384-dimensional vector as list of floats
        """
        if not self.is_ready():
            raise RuntimeError("Embedding service not initialized")

        if not text or not text.strip():
            raise ValueError("Text cannot be empty")

        try:
            # Run embedding generation in thread pool
            loop = asyncio.get_event_loop()
            embedding = await loop.run_in_executor(
                None,
                lambda: self.model.encode(text, convert_to_numpy=True)
            )

            # Convert numpy array to list
            return embedding.tolist()

        except Exception as e:
            logger.error(f"Failed to generate embedding: {str(e)}")
            raise

    def build_expert_text(self, expert_data: Dict[str, Any]) -> str:
        """
        Build comprehensive text representation of an expert for embedding
        """
        parts = []

        # Basic info
        if expert_data.get('name'):
            parts.append(f"Name: {expert_data['name']}")

        # Bio/Experience
        if expert_data.get('bio') or expert_data.get('experience_summary'):
            bio = expert_data.get('bio') or expert_data.get('experience_summary', '')
            parts.append(f"Experience: {bio}")

        # Skills
        if expert_data.get('skills'):
            skills = expert_data['skills']
            if isinstance(skills, list):
                skills_text = ", ".join(skills)
            else:
                skills_text = str(skills)
            parts.append(f"Skills: {skills_text}")

        # Domains/Expertise areas
        if expert_data.get('domains'):
            domains = expert_data['domains']
            if isinstance(domains, list):
                domains_text = ", ".join(domains)
            else:
                domains_text = str(domains)
            parts.append(f"Domains: {domains_text}")

        if expert_data.get('expertise_areas'):
            expertise = expert_data['expertise_areas']
            if isinstance(expertise, list):
                expertise_text = ", ".join(expertise)
            else:
                expertise_text = str(expertise)
            parts.append(f"Expertise Areas: {expertise_text}")

        # Patents, papers, products
        if expert_data.get('patents'):
            parts.append(f"Patents: {expert_data['patents']}")

        if expert_data.get('papers'):
            parts.append(f"Publications: {expert_data['papers']}")

        if expert_data.get('products'):
            parts.append(f"Products: {expert_data['products']}")

        # Combine all parts
        return " | ".join(parts)

    async def calculate_similarity(self, embedding1: List[float], embedding2: List[float]) -> float:
        """
        Calculate cosine similarity between two embeddings
        """
        try:
            vec1 = np.array(embedding1)
            vec2 = np.array(embedding2)

            # Cosine similarity
            dot_product = np.dot(vec1, vec2)
            norm1 = np.linalg.norm(vec1)
            norm2 = np.linalg.norm(vec2)

            if norm1 == 0 or norm2 == 0:
                return 0.0

            similarity = dot_product / (norm1 * norm2)
            return float(similarity)

        except Exception as e:
            logger.error(f"Failed to calculate similarity: {str(e)}")
            return 0.0

File: deeptech_backend/services/embeddingService.js
/**
 * Embedding Service
 * Generates semantic embeddings for text using sentence transformers
 * Model: all-MiniLM-L6-v2 (384 dimensions, fast, good quality)
 */

import { pipeline } from '@xenova/transformers';

class EmbeddingService {
  constructor() {
    this.model = null;
    this.modelName = 'Xenova/all-MiniLM-L6-v2';
    this.isInitializing = false;
    this.initializationPromise = null;
  }

  /**
   * Initialize the embedding model
   * Downloads model on first run, then caches locally
   */
  async initialize() {
    if (this.model) {
      return this.model;
    }

    // Prevent multiple simultaneous initializations
    if (this.isInitializing) {
      return this.initializationPromise;
    }

    this.isInitializing = true;
    
    try {
      console.log(`[EmbeddingService] Loading model: ${this.modelName}...`);
      this.initializationPromise = pipeline('feature-extraction', this.modelName);
      this.model = await this.initializationPromise;
      console.log('[EmbeddingService] Model loaded successfully');
      return this.model;
    } catch (error) {
      console.error('[EmbeddingService] Failed to load model:', error);
      throw error;
    } finally {
      this.isInitializing = false;
    }
  }

  /**
   * Preprocess text for embedding generation
   * @param {string} text - Raw text to preprocess
   * @returns {string} - Cleaned text
   */
  preprocessText(text) {
    if (!text) return '';
    
    return text
      .trim()
      .replace(/\s+/g, ' ') // Normalize whitespace
      .slice(0, 512); // Limit to 512 chars (model max)
  }

  /**
   * Generate embedding for a single text
   * @param {string} text - Text to embed
   * @returns {Promise<number[]>} - Embedding vector (384 dimensions)
   */
  async generateEmbedding(text) {
    await this.initialize();

    const cleanText = this.preprocessText(text);
    if (!cleanText) {
      throw new Error('Cannot generate embedding for empty text');
    }

    try {
      const output = await this.model(cleanText, {
        pooling: 'mean',
        normalize: true,
      });

      // Convert tensor to array
      const embedding = Array.from(output.data);
      
      return embedding;
    } catch (error) {
      console.error('[EmbeddingService] Error generating embedding:', error);
      throw error;
    }
  }

  /**
   * Generate embeddings for multiple texts (batch processing)
   * More efficient than calling generateEmbedding multiple times
   * @param {string[]} texts - Array of texts to embed
   * @returns {Promise<number[][]>} - Array of embedding vectors
   */
  async batchGenerateEmbeddings(texts) {
    await this.initialize();

    const cleanTexts = texts.map(t => this.preprocessText(t));
    
    try {
      const embeddings = await Promise.all(
        cleanTexts.map(text => this.generateEmbedding(text))
      );
      
      return embeddings;
    } catch (error) {
      console.error('[EmbeddingService] Error in batch generation:', error);
      throw error;
    }
  }

  /**
   * Calculate cosine similarity between two embeddings
   * @param {number[]} embedding1 - First embedding vector
   * @param {number[]} embedding2 - Second embedding vector
   * @returns {number} - Similarity score (0-1, higher is more similar)
   */
  cosineSimilarity(embedding1, embedding2) {
    if (embedding1.length !== embedding2.length) {
      throw new Error('Embeddings must have same dimensions');
    }

    let dotProduct = 0;
    let mag1 = 0;
    let mag2 = 0;

    for (let i = 0; i < embedding1.length; i++) {
      dotProduct += embedding1[i] * embedding2[i];
      mag1 += embedding1[i] * embedding1[i];
      mag2 += embedding2[i] * embedding2[i];
    }

    const magnitude = Math.sqrt(mag1) * Math.sqrt(mag2);
    
    if (magnitude === 0) return 0;
    
    return dotProduct / magnitude;
  }

  /**
   * Build text representation from expert profile
   * @param {object} expert - Expert object from database
   * @returns {string} - Combined text for embedding
   */
  buildExpertText(expert) {
    const parts = [];

    if (expert.bio) parts.push(expert.bio);
    
    if (expert.skills && Array.isArray(expert.skills)) {
      parts.push(expert.skills.join(' '));
    }
    
    if (expert.domains && Array.isArray(expert.domains)) {
      parts.push(expert.domains.join(' '));
    }

    if (expert.expertise_areas) {
      parts.push(expert.expertise_areas);
    }

    return parts.join('. ');
  }

  /**
   * Build text representation from project
   * @param {object} project - Project object from database
   * @returns {string} - Combined text for embedding
   */
  buildProjectText(project) {
    const parts = [];

    if (project.title) parts.push(project.title);
    if (project.description) parts.push(project.description);
    if (project.expected_outcome) parts.push(project.expected_outcome);
    
    if (project.domains && Array.isArray(project.domains)) {
      parts.push(project.domains.join(' '));
    }

    if (project.risk_categories && Array.isArray(project.risk_categories)) {
      parts.push(project.risk_categories.join(' '));
    }

    return parts.join('. ');
  }
}

// Export singleton instance
const embeddingService = new EmbeddingService();
export default embeddingService;
